drivers/video/gc2145.c

```diff
@@ -901,28 +905,65 @@ static int gc2145_set_output_format(const struct device *dev, int output_format)
 	return 0;
 }
 
-static int gc2145_set_resolution(const struct device *dev, uint16_t img_width, uint16_t img_height)
+static int gc2145_set_resolution(const struct device *dev, enum resolutions res)
 {
 	int ret;
 	const struct gc2145_config *cfg = dev->config;
 
-	uint16_t win_w = UXGA_HSIZE + 16;
-	uint16_t win_h = UXGA_VSIZE + 8;
-	uint16_t c_ratio = 1;
-	uint16_t r_ratio = 1;
-	uint16_t x = (((win_w / c_ratio) - UXGA_HSIZE) / 2);
-	uint16_t y = (((win_h / r_ratio) - UXGA_VSIZE) / 2);
-	uint16_t win_x = ((UXGA_HSIZE - win_w) / 2);
-	uint16_t win_y = ((UXGA_VSIZE - win_h) / 2);
+	uint16_t w = fmts[res].width_min;
+	uint16_t h = fmts[res].height_min;
+	uint16_t win_w;
+	uint16_t win_h;
+	uint16_t c_ratio;
+	uint16_t r_ratio;
+	uint16_t x;
+	uint16_t y;
+	uint16_t win_x;
+	uint16_t win_y;
+
+	if (res >= RESOLUTIONS_MAX) {
+		return -EIO;
+	}
+
+	w = fmts[res].width_min;
+	h = fmts[res].height_min;
+
+	/* Add the subsampling factor depending on resolution */
+	switch (res) {
+	case QVGA_RESOLUTION:
+		win_w = w * 3;
+		win_h = h * 3;
+		break;
+	case VGA_RESOLUTION:
+		win_w = w * 2;
+		win_h = h * 2;
+		break;
+	case UXGA_RESOLUTION:
+		win_w = w;
+		win_h = h;
+		break;
+	default:
+		LOG_ERR("Unsupported resolution %d", res);
+		return -EIO;
+	};
+
+	/* Calculates the window boundaries to obtain the desired resolution */
+	c_ratio = win_w / w;
+	r_ratio = win_h / h;
+	x = (((win_w / c_ratio) - w) / 2);
+	y = (((win_h / r_ratio) - h) / 2);
```

Comment:
IIUC the cases set `win_w` by multiplying `w` with a hardcoded number. Then, just after the `switch`, `c_ratio` values are calculated by dividing `win_w` by `w`. That's the hardcoded value.. :slightly_smiling_face: 
But this whole block looks fishy to me: `x` and `y` end up always zero (`w` - `w`).  
I suspect this logic originally dealt with rounding issues, at a time when `win_w` was unrelated to `w`. Or I am missing something...

Reply:
Got your point, it seems I'm looking too often to this code that I can't see more the obvious, just fixed.

On the other hand, some of the logic is also imported for the zoom handling, on the Arducam driver, the set resolution original takes also the zoom factor as variable, in that sense I'm discarding it since there is no zoom, much like the original code does, and I also removed the zoom logic, bacause it will not be supported yet (I need to check if zoom could be supported as standard video4zephyr control or if the API would need some kind of extension.

Please let me know if the current fix is enough for now.